lib/segment/src/index/field_index/binary_index.rs

```diff
@@ -0,0 +1,491 @@
+            falses: BitVec::new(),
+        }
+    }
+
+    pub fn get(&self, id: PointOffsetType) -> BinaryItem {
+        debug_assert!(self.trues.len() == self.falses.len());
+        if (id as usize) < self.trues.len() {
+            unsafe {
+                // SAFETY: we just checked that the id is within bounds
+                match (
+                    self.trues.get_unchecked(id as usize).as_ref(),
+                    self.falses.get_unchecked(id as usize).as_ref(),
+                ) {
+                    (true, true) => BinaryItem::Both,
+                    (true, false) => BinaryItem::True,
+                    (false, true) => BinaryItem::False,
+                    (false, false) => BinaryItem::None,
+                }
+            }
+        } else {
+            BinaryItem::None
+        }
+    }
+
+    pub fn set_or_insert(&mut self, id: PointOffsetType, has_true: bool, has_false: bool) {
+        if !has_true && !has_false {
+            return;
+        }
+
+        if (id as usize) >= self.trues.len() {
+            self.trues.resize(id as usize + 1, false);
+            self.falses.resize(id as usize + 1, false);
+        }
+
+        debug_assert!(self.trues.len() == self.falses.len());
+
+        unsafe {
+            // SAFETY: we just resized the vectors to be at least as long as the id
+            self.trues.set_unchecked(id as usize, has_true);
+            self.falses.set_unchecked(id as usize, has_false);
+        }
+    }
+
+    /// Removes the point from the index and tries to shrink the vectors if possible. If the index is not within bounds, does nothing
+    pub fn remove(&mut self, id: PointOffsetType) {
+        if (id as usize) < self.trues.len() {
+            self.trues.set(id as usize, false);
+            self.falses.set(id as usize, false);
+        }
+
+        // TODO: should we avoid shrinking the vecs on every remove?
+        self.shrink();
+    }
+
+    /// Shrinks the vectors to the last populated index
+    fn shrink(&mut self) {
+        let last_populated_index = self.trues.last_one().max(self.falses.last_one());
+        match last_populated_index {
+            Some(index) if index < self.trues.len() - 1 => {
+                self.trues.truncate(index + 1);
+                self.falses.truncate(index + 1);
+            }
+            None => {
+                self.trues.clear();
+                self.falses.clear();
+            }
+            _ => {}
+        }
+    }
+
+    pub fn count_trues(&self) -> usize {
+        self.trues.count_ones()
+    }
+
+    pub fn count_falses(&self) -> usize {
+        self.falses.count_ones()
+    }
+
+    pub fn indexed_count(&self) -> usize {
+        self.trues.count_ones().max(self.falses.count_ones())
+    }
+
+    pub fn iter(&self) -> BinaryMemoryIterator {
+        let last_false = self.falses.last_one();
+        let last_true = self.trues.last_one();
+        let end = last_false.max(last_true).unwrap_or(0) + 1;
```

Comment:
It's true... the default value should be -1 so that the -1 +1 = 0, because we are looking for the length of the memory, and `last_one()` gives us index. Though I like your solution of just iterating over the whole vecs. We already try to keep it as shrunk as possible.